app/controllers/auth_controller.rb

```diff
@@ -26,10 +29,20 @@ def token
         token_type: 'Bearer',
         expires_in: time_to_live,
         scope: scope,
-        patient: Patient.all.first.id
+        patient: Patient.first&.id
       }
+    elsif params.key?(:code) && params.key?(:client_id)
```

Comment:
Should a correct `client_id` with an incorrect `code` return an `invalid_grant` instead of an `invalid_client`?

https://datatracker.ietf.org/doc/html/rfc6749#section-5.2

Reply:
Yes, I'm adding that case

Reply:
I'm also adding a check for `grant_type` param; if `grant_type` **is** included but does not equal `authorization_code` I return `invalid_grant_type`

OAuth 2 actually requires `grant_type` but I'm keeping it optional: <https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.3>

Reply:
I think we should add a test for that case, if you're adding the `elsif` statement that checks for an invalid grant type.

Reply:
And is there a reason you're keeping it optional? it seems like grant_type is required according to [7.1.3](http://hl7.org/fhir/smart-app-launch/#step-3-app-exchanges-authorization-code-for-access-token)

Reply:
I mean I'm keeping the param grant_type=authorization_code optional, since in the FHIR spec grant_type is not listed as a required param

Reply:
I don't know where you're seeing that `grant_type` isn't required. Mick linked to the section that says it is required.
![Screen Shot 2021-06-21 at 1 40 20 PM](https://user-images.githubusercontent.com/15969967/122804919-71bbea00-d296-11eb-9eba-f777ccdbe076.png)

Reply:
Ah its only for the /auth/token action. My bad I'm fixing this.